### PR TITLE
Add setting to toggle language file usage between tsv and yml and allow the "factionless" placeholder to be edited.

### LINKED ISSUE #1426: "Factionless" placeholder - editable
**Is your feature request related to a problem? Please describe.**
I just want, for everyone, to have an permission for editing this placeholder.

**Describe the solution you'd like**
Clean command, or option in the config for editting the color, and text of this placeholder.

**Describe alternatives you've considered**
I dont have one. 

**Additional context**
I think this feature is pretty important for everyone, who wants to create some SMP server. 😄

### PR DESCRIPTION
**Pull request by VoChiDanh**
closes #1426

### CODE REVIEW

File: src/main/java/dansplugins/factionsystem/integrators/FiefsIntegrator.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/integrators/FiefsIntegrator.java:19]: What's the purpose of this change?

File: src/main/java/dansplugins/factionsystem/placeholders/PlaceholderAPI.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/placeholders/PlaceholderAPI.java:66]: We should get the config service via dependency injection, not create a new instance each time we need it

File: src/main/java/dansplugins/factionsystem/services/MessageService.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/services/MessageService.java:14]: None of the fields or methods in this class need to be static

File: src/main/java/dansplugins/factionsystem/services/PlayerService.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/services/PlayerService.java:20]: Variables should start with a lower case letter. replace can be a primitive boolean, it doesn't need the wrapper class, which allows null values to be passed in

File: src/main/java/dansplugins/factionsystem/services/PlayerService.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/services/PlayerService.java:11]: None of the methods in this class need to be static.

**Review by alyphen — CHANGES_REQUESTED** [review]

File: src/main/java/dansplugins/factionsystem/integrators/FiefsIntegrator.java
Comment (VoChiDanh) [reply src/main/java/dansplugins/factionsystem/integrators/FiefsIntegrator.java:19]: Looks like I made the wrong change

File: src/main/java/dansplugins/factionsystem/placeholders/PlaceholderAPI.java
Comment (VoChiDanh) [reply src/main/java/dansplugins/factionsystem/placeholders/PlaceholderAPI.java]: This is probably the fastest and best way for users

File: src/main/java/dansplugins/factionsystem/services/PlayerService.java
Comment (VoChiDanh) [reply src/main/java/dansplugins/factionsystem/services/PlayerService.java]: replace here means, if true, it will get the message in new message, and if false then just type the first one of the message (I don't know what to call it but it's like getText(), I just type the one inside the quotes) then the method takes that line in the language file of yml to send it to the player, making typing a bit shorter

File: src/main/java/dansplugins/factionsystem/commands/MembersCommand.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/commands/MembersCommand.java:81]: We don't want to create a new instance of the main plugin class - pass the reference into the command.

File: src/main/java/dansplugins/factionsystem/integrators/CurrenciesIntegrator.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/integrators/CurrenciesIntegrator.java:15]: This inverts the logic! Surely if we want to check currencies _is_ present this should be returning that it isn't null?

File: src/main/java/dansplugins/factionsystem/integrators/DynmapIntegrator.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/integrators/DynmapIntegrator.java:404]: Appreciate that the original code wasn't yours but since we're changing code in the area, perhaps we could use the plugin's logger rather than print statements?

File: src/main/java/dansplugins/factionsystem/objects/helper/FactionFlags.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/objects/helper/FactionFlags.java:150]: This, combined with the currencies integrator change, makes this more difficult to follow. We should be returning if currencies is not present, as the flag isn't supported.

File: src/main/java/dansplugins/factionsystem/services/CommandService.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/services/CommandService.java:115]: We probably want `replace` rather than `replaceAll` here - `replaceAll` does a regex replacement. Somewhat confusingly, `replace` in Java will replace all instances of a string.

File: src/main/java/dansplugins/factionsystem/services/ConfigService.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/services/ConfigService.java:170]: Perhaps `getConfig().set(...)` would be better here, or change the check to be `contains` rather than `isString`.
`addDefault` in combination with `options().copyDefaults()` will work if the key is not set, but that's not what we're checking - we're checking that the key _isn't a string_ - so in the case some other type is there, this will do nothing at all (as the default has already been overridden)

File: src/main/java/dansplugins/factionsystem/services/MessageService.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/services/MessageService.java:18]: Pass the plugin instance into the message service.

File: src/main/java/dansplugins/factionsystem/services/MessageService.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/services/MessageService.java:25]: Perhaps use the plugin's logger to log exceptions.

File: src/main/java/dansplugins/factionsystem/services/PlayerService.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/services/PlayerService.java:12]: Variables should start with lowercase (i.e. `oldType` and `newType`)

**Review by alyphen — CHANGES_REQUESTED** [review]

File: src/main/java/dansplugins/factionsystem/commands/AddLawCommand.java
Comment (VoChiDanh) [reply src/main/java/dansplugins/factionsystem/commands/AddLawCommand.java:50]: What is dependency injection mean?

File: src/main/java/dansplugins/factionsystem/integrators/CurrenciesIntegrator.java
Comment (VoChiDanh) [reply src/main/java/dansplugins/factionsystem/integrators/CurrenciesIntegrator.java:15]: It will return false if plugin installed

File: src/main/java/dansplugins/factionsystem/integrators/DynmapIntegrator.java
Comment (VoChiDanh) [reply src/main/java/dansplugins/factionsystem/integrators/DynmapIntegrator.java:404]: I dont think i changed here, i think it automatic changed by reformat code

File: src/main/java/dansplugins/factionsystem/services/ConfigService.java
Comment (VoChiDanh) [reply src/main/java/dansplugins/factionsystem/services/ConfigService.java:170]: i follow the methods above, maybe i should change the whole method in there?

File: src/main/java/dansplugins/factionsystem/services/PlayerService.java
Comment (VoChiDanh) [reply src/main/java/dansplugins/factionsystem/services/PlayerService.java:12]: Oki, i'll change it in next push

File: src/main/java/dansplugins/factionsystem/commands/AddLawCommand.java
Comment (dmccoystephenson) [reply src/main/java/dansplugins/factionsystem/commands/AddLawCommand.java:50]: If my understanding is correct, dependency injection is where any objects that a class depends on is passed through the constructor as parameters.

for example say that class A depends on class B - this is what class A would look like:

```
class A {
    b = null;

    A(B b) {
        this.b = b;
    }
}
```

Comment (alyphen) [reply src/main/java/dansplugins/factionsystem/commands/AddLawCommand.java:50]: Dependency injection is a software design pattern in which the dependencies of a given service or component are passed into that service or component.
This means that services and components do not need to know the implementation details of each dependency, but is given everything it needs to function when it is created.
By building software in this way, we end up with loosely coupled components, which makes an application more flexible and easier to change parts later.

File: src/main/java/dansplugins/factionsystem/services/ConfigService.java
Comment (dmccoystephenson) [reply src/main/java/dansplugins/factionsystem/services/ConfigService.java:170]: This is an old habit of mine. We can replace any instances of .addDefault() with .set(), that's what I do with more recent projects now.

File: src/main/java/dansplugins/factionsystem/services/MessageService.java
Comment (dmccoystephenson) [reply src/main/java/dansplugins/factionsystem/services/MessageService.java:18]: Yeah pass the main class to MessageService.java rather than instantiating a new instance here.

File: src/main/java/dansplugins/factionsystem/commands/AddLawCommand.java
Comment (VoChiDanh) [reply src/main/java/dansplugins/factionsystem/commands/AddLawCommand.java:50]: oh I see. I'm not used to that style, I'll try anyway

File: src/main/java/dansplugins/factionsystem/integrators/CurrenciesIntegrator.java
Comment (alyphen) [reply src/main/java/dansplugins/factionsystem/integrators/CurrenciesIntegrator.java:15]: That runs counter to the method's name: `isCurrenciesPresent` should return whether currencies is present, not missing.
Either the change in behaviour should be reverted or the method should be renamed to make the intent clear.

Comment (VoChiDanh) [reply src/main/java/dansplugins/factionsystem/integrators/CurrenciesIntegrator.java:15]: oki, i'll change

**Comment by dmccoystephenson** [comment]
closes #1426

File: src/main/java/dansplugins/factionsystem/commands/DuelCommand.java
Comment (dmccoystephenson) [inline src/main/java/dansplugins/factionsystem/commands/DuelCommand.java:46]: Nice catch, thanks!

File: src/main/resources/en-us.tsv
Comment (dmccoystephenson) [inline src/main/resources/en-us.tsv:1]: Do language key lookups still work with the added spaces here?

**Review by dmccoystephenson — APPROVED** [review]
This looks good to me! Thank you very much for your contribution. I'll let @alyphen take a final look before we merge this in.

File: src/main/resources/en-us.tsv
Comment (VoChiDanh) [reply src/main/resources/en-us.tsv:1]: yes

File: src/main/java/dansplugins/factionsystem/commands/AllyCommand.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/commands/AllyCommand.java:112]: use `replace` over `replaceAll` (`replaceAll` is for regex)

File: src/main/java/dansplugins/factionsystem/commands/CheckClaimCommand.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/commands/CheckClaimCommand.java:49]: Use `replace` over `replaceAll`

File: src/main/java/dansplugins/factionsystem/commands/CreateCommand.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/commands/CreateCommand.java:102]: Use `replace` over `replaceAll`

File: src/main/java/dansplugins/factionsystem/commands/DeclareWarCommand.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/commands/DeclareWarCommand.java:140]: Use `replace` over `replaceAll`

File: src/main/java/dansplugins/factionsystem/commands/DisbandCommand.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/commands/DisbandCommand.java:65]: Pass in the plugin instance rather than constructing a new one

File: src/main/java/dansplugins/factionsystem/commands/DisbandCommand.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/commands/DisbandCommand.java:87]: Use `replace` over `replaceAll`

File: src/main/java/dansplugins/factionsystem/commands/GateCommand.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/commands/GateCommand.java:49]: Pass in the plugin instance rather than constructing a new one

File: src/main/java/dansplugins/factionsystem/commands/GrantIndependenceCommand.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/commands/GrantIndependenceCommand.java:62]: Use replace over replaceAll

File: src/main/java/dansplugins/factionsystem/commands/GrantIndependenceCommand.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/commands/GrantIndependenceCommand.java:66]: Use replace over replaceAll

File: src/main/java/dansplugins/factionsystem/commands/InviteCommand.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/commands/InviteCommand.java:94]: Use replace over replaceAll

File: src/main/java/dansplugins/factionsystem/commands/InviteCommand.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/commands/InviteCommand.java:108]: Use replace over replaceAll

File: src/main/java/dansplugins/factionsystem/commands/InvokeCommand.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/commands/InvokeCommand.java:86]: Use replace over replaceAll

File: src/main/java/dansplugins/factionsystem/commands/InvokeCommand.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/commands/InvokeCommand.java:89]: Use replace over replaceAll

File: src/main/java/dansplugins/factionsystem/commands/InvokeCommand.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/commands/InvokeCommand.java:92]: Use replace over replaceAll

File: src/main/java/dansplugins/factionsystem/commands/StatsCommand.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/commands/StatsCommand.java:37]: Use plugin instance rather than creating a new one

**Review by alyphen — CHANGES_REQUESTED** [review]
Services are looking much better, awesome work.

There's two minor flaws I can see in this still: usage of replaceAll where replace should be used (replaceAll should only be used for regex replacements) and creating new instances of the plugin class (this should never be done, pass the instance of the main class around rather than creating new ones)

I haven't commented on every single one since it's the same issue but I'm confident you can find them.

File: src/main/java/dansplugins/factionsystem/commands/AllyCommand.java
Comment (VoChiDanh) [reply src/main/java/dansplugins/factionsystem/commands/AllyCommand.java:112]: I use replaceAll because that can use number but Replace can't, however replaceAll still is better. This cannot be forced

**Comment by VoChiDanh** [comment]
My machine just had a problem so maybe tomorrow morning I can double check if I'm using the instance somewhere else. Besides Using replace with replaceall, they are the same but replaceall can convert all to string and use it, and replace it is almost only used to change characters such as special characters... I will still keep as replaceall for my safety

**Comment by alyphen** [comment]
The only difference between replace and replaceAll is that replaceAll uses a regular expression while replace uses plain strings.

File: src/main/java/dansplugins/factionsystem/commands/AllyCommand.java
Comment (dmccoystephenson) [reply src/main/java/dansplugins/factionsystem/commands/AllyCommand.java:112]: Try using replace and seeing if the behavior of the program changes.

**Comment by dmccoystephenson** [comment]
> The only difference between replace and replaceAll is that replaceAll uses a regular expression while replace uses plain strings.

Could you clarify why it's preferable to replace plain strings rather than using a regular expression here?

File: src/main/java/dansplugins/factionsystem/commands/DisbandCommand.java
Comment (dmccoystephenson) [reply src/main/java/dansplugins/factionsystem/commands/DisbandCommand.java:65]: Seconded.

**Comment by alyphen** [comment]
Regular expressions are for pattern matching. When you want to match exact text you use a string, when you want to match a pattern like "two numbers followed by between 4 and 5 letters between a and f" or "any text surrounded by asterisks" or similar, that's when you use replaceAll (or Pattern if you need more advanced functionality)

File: src/main/java/dansplugins/factionsystem/commands/AllyCommand.java
Comment (dmccoystephenson) [reply src/main/java/dansplugins/factionsystem/commands/AllyCommand.java]: Any instances of .replaceAll have been replaced with .replace.

File: src/main/java/dansplugins/factionsystem/commands/DisbandCommand.java
Comment (dmccoystephenson) [reply src/main/java/dansplugins/factionsystem/commands/DisbandCommand.java]: Done.

**Comment by dmccoystephenson** [comment]
This is ready for re-review @alyphen

File: src/main/java/dansplugins/factionsystem/MedievalFactions.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/MedievalFactions.java:64]: This method isn't necessary

File: src/main/java/dansplugins/factionsystem/MedievalFactions.java
Comment (alyphen) [inline src/main/java/dansplugins/factionsystem/MedievalFactions.java:77]: This isn't necessary

**Review by alyphen — CHANGES_REQUESTED** [review]

File: src/main/java/dansplugins/factionsystem/MedievalFactions.java
Comment (dmccoystephenson) [reply src/main/java/dansplugins/factionsystem/MedievalFactions.java]: This has been removed.

File: src/main/java/dansplugins/factionsystem/MedievalFactions.java
Comment (dmccoystephenson) [reply src/main/java/dansplugins/factionsystem/MedievalFactions.java]: This has been removed.

**Comment by dmccoystephenson** [comment]
This should be ready for re-review. @alyphen

**Review by alyphen — APPROVED** [review]

**Pull request merged**
